feat: Add documentation for mutually recursive data types (#227)
grain-lang/grain-lang.org: src/guide/data_types.md

print(counter) // { count: 1 }
```

## Mutually Recursive Data Types

> In some instances, it is necessary to define multiple data types which refer to one another. For example, imagine you wanted to define your own number system based on [Peano Arithmetic](https://en.wikipedia.org/wiki/Peano_axioms). You might try something like this:

Reviewer: I dunno why, but reading "one another" made me think that I couldn't do something like defining a type and then another type that refers to the first one. I think this makes it more explicit what we're talking about here:
```suggestion
In some instances, it's necessary to define multiple data types where each of the types refers to the others. For example, imagine you wanted to define your own number system based on [Peano Arithmetic](https://en.wikipedia.org/wiki/Peano_axioms). You might try something like this:
```

Author: Maybe `which refer to each other` would be more clear?

Reviewer: I think that would be fine 👍

```grain
enum EvenNumber {
  Zero, // 0
  OddPlusOne(OddNumber) // one plus an odd number
}

enum OddNumber {
  EvenPlusOne(EvenNumber) // one plus an even number
}

let two = OddPlusOne(EvenPlusOne(Zero)) // 2 == (0 + 1) + 1
```

> If you try to run this program, it won't work! That's because, when the Grain compiler reads the definition for `EvenNumber`, it sees the reference to `OddNumber` before it's read its definition. This means that the compiler isn't sure what `OddNumber` is referring to! Luckily, there is a simple fix for this: we can place a comma after `EvenNumber`'s definition to tell the compiler that the next data definition is allowed to be recursive with `EvenNumber`:

Reviewer: ```suggestion
If you try to run this program, it won't work! That's because when the Grain compiler reads the definition for `EvenNumber`, it sees the reference to `OddNumber` before it's read its definition. This means that the compiler isn't sure what `OddNumber` is referring to! Luckily, there is a simple fix for this: we can place a comma after `EvenNumber`'s definition to tell the compiler that the next data definition is allowed to be recursive with `EvenNumber`:
```

Author: The `when` clause here is essential to the sentence's meaning, so there should not be a comma here. See [#3 on this page](https://owl.purdue.edu/owl/general_writing/punctuation/commas/extended_rules_for_commas.html).
